File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        hm = {}
        maxrepeat = 0

        l = r = 0
        maxf = 0
        while r < len(s):
            hm[s[r]] = 1 + hm.get(s[r], 0)
            maxf = max(maxf, hm[s[r]])
            
            while (r - l + 1) - maxf > k:
                hm[s[l]] -= 1
                l += 1

            maxrepeat = max(maxrepeat, r - l + 1)
            r += 1

        return maxrepeat

File: Data Structures & Algorithms/max-water-container/submission-3.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    maxArea(heights) {
        let ret = 0
        
        let l = 0
        let r = heights.length - 1


        while (l < r) {
            const h = Math.min(heights[l], heights[r])

            ret = Math.max(h * (r - l), ret)

            if (heights[l] > heights[r]) {
                r--
            } else {
                l++
            }
        }

        return ret
    }
}
